File: src/providers/ClaudeVoiceProtocol.cpp
#include "providers/ClaudeVoiceProtocol.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QSet>

namespace speecher {
namespace {

constexpr qsizetype kMaximumKeytermBytes = 1024;

QString redactedErrorSummary(const QJsonObject &object)
{
    const QJsonValue errorValue = object.value(QStringLiteral("error"));
    QJsonObject error = errorValue.toObject();
    if (error.isEmpty() && errorValue.isString()) {
        return errorValue.toString().left(240);
    }
    QStringList parts;
    for (const QString &key : {QStringLiteral("type"), QStringLiteral("code"), QStringLiteral("error_code"), QStringLiteral("message"), QStringLiteral("description")}) {
        const QString value = error.value(key).toString(object.value(key).toString());
        if (!value.isEmpty()) {
            parts << key + QStringLiteral("=") + value.left(240);
        }
    }
    return parts.join(QStringLiteral(" "));
}

bool envFlag(const char *name, bool defaultValue)
{
    const QString value = qEnvironmentVariable(name).trimmed().toLower();
    if (value.isEmpty()) {
        return defaultValue;
    }
    if (value == QStringLiteral("1") || value == QStringLiteral("true") || value == QStringLiteral("yes") || value == QStringLiteral("on")) {
        return true;
    }
    if (value == QStringLiteral("0") || value == QStringLiteral("false") || value == QStringLiteral("no") || value == QStringLiteral("off")) {
        return false;
    }
    return defaultValue;
}

bool typedInterimsEnabled()
{
    if (envFlag("CLAUDE_CODE_VOICE_FORWARD_INTERIMS_TYPED", false)) {
        return true;
    }
    return envFlag("SPEECHER_CLAUDE_FORWARD_INTERIMS_TYPED", true);
}

} // namespace

ClaudeVoiceEvent parseClaudeVoiceEvent(const QString &message)
{
    const QJsonObject object = QJsonDocument::fromJson(message.toUtf8()).object();
    const QString type = object.value(QStringLiteral("type")).toString();
    if (type == QStringLiteral("TranscriptInterim") || type == QStringLiteral("TranscriptText")) {
        return {ClaudeVoiceEventKind::Working, object.value(QStringLiteral("data")).toString(), {}};
    }
    if (type == QStringLiteral("TranscriptEndpoint")) {
        return {ClaudeVoiceEventKind::Endpoint, object.value(QStringLiteral("data")).toString(), {}};
    }
    if (type == QStringLiteral("TranscriptError")) {
        return {ClaudeVoiceEventKind::TranscriptError, {}, redactedErrorSummary(object)};
    }
    if (type == QStringLiteral("error") || object.contains(QStringLiteral("error"))) {
        return {ClaudeVoiceEventKind::Error, {}, redactedErrorSummary(object)};
    }
    return {};
}

QUrlQuery claudeVoiceStreamQuery(const QStringList &vocabulary)
{
    Q_UNUSED(vocabulary)
    QUrlQuery query;
    query.addQueryItem(QStringLiteral("encoding"), QStringLiteral("linear16"));
    query.addQueryItem(QStringLiteral("sample_rate"), QStringLiteral("16000"));
    query.addQueryItem(QStringLiteral("channels"), QStringLiteral("1"));
    query.addQueryItem(QStringLiteral("endpointing_ms"), QStringLiteral("300"));
    query.addQueryItem(QStringLiteral("utterance_end_ms"), QStringLiteral("1000"));
    query.addQueryItem(QStringLiteral("language"), QStringLiteral("en"));
    query.addQueryItem(QStringLiteral("use_conversation_engine"), QStringLiteral("true"));
    if (typedInterimsEnabled()) {
        query.addQueryItem(QStringLiteral("forward_interims"), QStringLiteral("typed"));
    }
    query.addQueryItem(QStringLiteral("stt_provider"), QStringLiteral("deepgram-nova3"));
    return query;
}

QByteArray claudeVoiceKeytermsHeader(const QStringList &vocabulary)
{
    QByteArray header;
    QSet<QByteArray> seen;
    for (const QString &value : vocabulary) {
        const QString simplified = value.simplified();
        const QByteArray term = simplified.toLatin1();
        const QByteArray key = term.toLower();
        if (term.isEmpty()
            || QString::fromLatin1(term) != simplified
            || seen.contains(key)) {
            continue;
        }
        const qsizetype separatorBytes = header.isEmpty() ? 0 : 1;
        if (header.size() + separatorBytes + term.size() > kMaximumKeytermBytes) {
            continue;
        }
        seen.insert(key);
        if (!header.isEmpty()) {
            header.append(',');
        }
        header.append(term);
    }
    return header;
}

} // namespace speecher

File: android/protocol/src/main/kotlin/app/speecher/protocol/ClaudeVoiceProtocol.kt
package app.speecher.protocol

import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

private const val MAX_KEYTERMS_BYTES = 1024
private const val MAX_ERROR_FIELD_CHARS = 240
private val ERROR_SUMMARY_KEYS = listOf("type", "code", "error_code", "message", "description")

/** A server message on the Claude voice stream, reduced to what the client acts on. */
sealed interface ClaudeVoiceEvent {
    /** An interim or partial transcript (`TranscriptInterim` or `TranscriptText`). */
    data class Working(val text: String) : ClaudeVoiceEvent

    /** The end of an utterance (`TranscriptEndpoint`); [text] may be empty. */
    data class Endpoint(val text: String) : ClaudeVoiceEvent

    /** The speech-to-text backend failed (`TranscriptError`). */
    data class TranscriptError(val summary: String) : ClaudeVoiceEvent

    /** Any other message of type `error` or carrying an `error` field. */
    data class ServerError(val summary: String) : ClaudeVoiceEvent

    /** Anything else, including text that is not a JSON object. */
    data object Unknown : ClaudeVoiceEvent
}

/**
 * Query parameters for the stream URL, in the order they are sent.
 *
 * `forward_interims=typed` is on by default. `SPEECHER_CLAUDE_FORWARD_INTERIMS_TYPED` can turn it
 * off, unless `CLAUDE_CODE_VOICE_FORWARD_INTERIMS_TYPED` forces it on.
 */
fun claudeVoiceStreamQuery(env: (String) -> String? = System::getenv): List<Pair<String, String>> =
    buildList {
        add("encoding" to "linear16")
        add("sample_rate" to "16000")
        add("channels" to "1")
        add("endpointing_ms" to "300")
        add("utterance_end_ms" to "1000")
        add("language" to "en")
        add("use_conversation_engine" to "true")
        if (typedInterimsEnabled(env)) add("forward_interims" to "typed")
        add("stt_provider" to "deepgram-nova3")
    }

/**
 * The `x-config-keyterms` header value: whitespace-collapsed terms joined by commas, first spelling
 * wins among case-insensitive duplicates. Terms with characters outside Latin-1 are dropped, as are
 * terms that would push the header past 1024 bytes; shorter terms after them can still fit.
 *
 * Every character is in U+0000..U+00FF, so the length is the byte count when sent as ISO-8859-1.
 */
fun claudeVoiceKeytermsHeader(vocabulary: Iterable<String>): String {
    val header = StringBuilder()
    val seen = mutableSetOf<String>()
    for (value in vocabulary) {
        val term = value.simplified()
        val key = term.lowercaseAscii()
        if (term.isEmpty() || term.any { it > 'ÿ' } || key in seen) continue
        val separator = if (header.isEmpty()) "" else ","
        if (header.length + separator.length + term.length > MAX_KEYTERMS_BYTES) continue
        seen += key
        header.append(separator).append(term)
    }
    return header.toString()
}

/**
 * Classifies one text frame from the stream. [ClaudeVoiceEvent.Working] and `Endpoint` text is
 * passed through untrimmed.
 */
fun parseClaudeVoiceEvent(message: String): ClaudeVoiceEvent {
    val json =
        try {
            Json.parseToJsonElement(message.removePrefix("\uFEFF")) as? JsonObject
        } catch (e: SerializationException) {
            null
        }
            ?: return ClaudeVoiceEvent
                .Unknown // Qt also treats malformed JSON and non-objects as an empty object

    val type = json.string("type")
    return when (type) {
        "TranscriptInterim",
        "TranscriptText" -> ClaudeVoiceEvent.Working(json.string("data").orEmpty())
        "TranscriptEndpoint" -> ClaudeVoiceEvent.Endpoint(json.string("data").orEmpty())
        "TranscriptError" -> ClaudeVoiceEvent.TranscriptError(json.redactedErrorSummary())
        else ->
            if (type == "error" || "error" in json) {
                ClaudeVoiceEvent.ServerError(json.redactedErrorSummary())
            } else {
                ClaudeVoiceEvent.Unknown
            }
    }
}

/**
 * Keeps only known descriptive fields, each cut to 240 chars. A field is read from the nested
 * `error` object when it holds a string there (even an empty one), otherwise from the top level.
 */
private fun JsonObject.redactedErrorSummary(): String {
    string("error")?.let {
        return it.take(MAX_ERROR_FIELD_CHARS)
    }
    val nested = this["error"] as? JsonObject
    return ERROR_SUMMARY_KEYS.mapNotNull { key ->
            val value = nested?.string(key) ?: string(key)
            if (value.isNullOrEmpty()) null else "$key=${value.take(MAX_ERROR_FIELD_CHARS)}"
        }
        .joinToString(" ")
}

/** The value at [key] if it is a JSON string; numbers, booleans and null don't count. */
private fun JsonObject.string(key: String): String? =
    (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun typedInterimsEnabled(env: (String) -> String?): Boolean =
    envFlag(env("CLAUDE_CODE_VOICE_FORWARD_INTERIMS_TYPED"), default = false) ||
        envFlag(env("SPEECHER_CLAUDE_FORWARD_INTERIMS_TYPED"), default = true)

private fun envFlag(value: String?, default: Boolean): Boolean =
    when (value?.trim(Char::isQtWhitespace)?.lowercase()) {
        "1",
        "true",
        "yes",
        "on" -> true
        "0",
        "false",
        "no",
        "off" -> false
        else -> default
    }

/** Trims, then replaces each internal whitespace run with one space. */
private fun String.simplified(): String = buildString {
    for (c in this@simplified.trim(Char::isQtWhitespace)) {
        when {
            !c.isQtWhitespace() -> append(c)
            last() != ' ' -> append(' ')
        }
    }
}

// QChar::isSpace(): unlike Char.isWhitespace(), it includes NEL (U+0085) and excludes
// U+001C..U+001F.
private fun Char.isQtWhitespace(): Boolean =
    this in '\t'..'\r' || this == '\u0085' || Character.isSpaceChar(this)

// Only ASCII letters fold, so "É" and "é" count as different terms.
private fun String.lowercaseAscii(): String = buildString {
    for (c in this@lowercaseAscii) append(if (c in 'A'..'Z') c.lowercaseChar() else c)
}
